parsers/US_ERCOT.py

```diff
@@ -0,0 +1,366 @@
+                        "zoneKey": zone_key,
+                        "datetime": hour_dt,
+                        "production": production,
+                        "storage": storage,
+                        "source": "ercot.com",
+                    }
+                )
+
+    return all_data_points
+
+
+def get_sheet_from_date(year: int, month: str, session: Session):
+    if year > 2020:
+        url = HISTORICAL_GENERATION_URL[str(year)]
+        return pd.read_excel(url, engine="openpyxl", sheet_name=month)
+    else:
+        url = HISTORICAL_GENERATION_URL["all_previous"]
+        response = session.get(url)
+
+        if response.content.startswith(b"PK"):
+            zip_data = BytesIO(response.content)
+        else:
+            try:
+                decompressed = gzip.decompress(response.content)
+                zip_data = BytesIO(decompressed)
+            except gzip.BadGzipFile as err:
+                raise ValueError("File is neither a ZIP nor a gzipped file") from err
+
+        year_file = f"IntGenbyFuel{year}.xlsx"
+
+        with zipfile.ZipFile(zip_data) as zf:
+            if year_file not in zf.namelist():
+                raise NotImplementedError(
+                    f"Data for year {year} not found in historical data"
+                )
+            with zf.open(year_file) as excel_file:
+                return pd.read_excel(excel_file, engine="openpyxl", sheet_name=month)
+
+
+def fetch_historical_production(
+    zone_key: ZoneKey,
+    session: Session,
+    target_datetime: datetime,
+    logger: Logger = getLogger(__name__),
+) -> list:
+    if target_datetime.tzinfo is None:
+        target_datetime = target_datetime.replace(tzinfo=timezone.utc)
+
+    end = target_datetime + timedelta(hours=1)
+    start = end - timedelta(days=1)
+    year = target_datetime.year
+    month = target_datetime.strftime("%b")
+
+    all_data_points = []
+
+    df = get_sheet_from_date(year, month, session)
+
+    df["Date"] = pd.to_datetime(df["Date"])
+    time_columns = df.columns[4:]
+    datapoints_by_date = {}
+
+    for _, row in df.iterrows():
+        date = datetime.strptime(str(row["Date"]), "%Y-%m-%d %H:%M:%S").replace(
+            tzinfo=TX_TZ
+        )
+        if date < (start + timedelta(days=-1)) or date > (end + timedelta(days=1)):
+            continue
+
+        production_source = GENERATION_MAPPING[row["Fuel"]]
+
+        for hour in range(0, 24):
+            hour_dt = date + timedelta(hours=hour)
+
+            if hour_dt < start or hour_dt > end:
+                continue
+
+            # We get the data in 15 minute intervals, so we need to sum the 4 columns that represent an hour
+            start_idx = hour * 4
+            end_idx = start_idx + 4
+            hour_cols = time_columns[start_idx:end_idx]
+            hour_value = row[hour_cols].sum()
+
+            if hour_dt not in datapoints_by_date:
+                datapoints_by_date[hour_dt] = {"storage": {}, "production": {}}
```

Comment:
You can avoid a lot of these manual checks by using the parser classes that handles these checks internally. (Check the ENTSOE parser for an example).

I'd recommend using them as they also do some internal validation to ensure all datapoints are valid in a uniform way.

Reply:
I don't think I can avoid all of them because the data is read for a whole day for one production mode and then afterwards a whole day for another production mode. So I need to temporarily store them before I can add the whole hour of production mix. If you can think of a better way to do it, let me know:)

But I will try to modify the code to use the parser classes as much as possible.

Reply:
Ah I see!

Maybe you can read all of them and zip them? That would ensure they are all the same length as well. (just throwing out an idea, I don't know if it will work in this case).

But please do use the parser classes as they will track negative corrections and automatically round values to 1W incements.

Reply:
Another alternative is to create 1 production / storage list for each mode and then merging all the production and storage. The parser classes have built in functions for that.

Reply:
True but I don't think its worth spending time on now, so I'll leave it as it is, unless you have a strong opinion about it.

Reply:
I'm not sure if the grid modeling team needs it otherwise I'm fine with it :)